[Frontend] Add a way to print features supported by the compiler

xedin opened the pull request:
- Replace `-emit-supported-features` with `-emit-supported-arguments` mode because "features" part was never implemented.
- Introduce `-print-supported-features` option that prints supported upcoming/experimental features and their metadata.

Resolves: rdar://148124973
<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Frontend] Add `-print-supported-features` option

This is a replacement for `-emit-supported-features` that prints
all of the upcoming/experimental features supported by the compiler
with some additional meta information in JSON format to stdout.

xedin commented:
https://github.com/swiftlang/swift-driver/pull/1875
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented on lib/Basic/SupportedFeatures.cpp:
```diff
@@ -0,0 +1,68 @@
+    out << " }";
+  };
+
+  out << "  \"SupportedFeatures\": {\n";
+  out << "    \"Upcoming\": [\n";
+  llvm::interleave(upcoming, printFeature, [&out] { out << ",\n"; });
+  out << "\n    ],\n";
+
+  out << "    \"Experimental\": [\n";
+  llvm::interleave(experimental, printFeature, [&out] { out << ",\n"; });
+  out << "\n    ]\n";
+
+  out << "  }\n";
+  out << "}\n";
```
Yeah, I was thinking maybe “feature_flags” maybe

xedin replied:
Let's just go with "features".

xedin commented on lib/Basic/SupportedFeatures.cpp:
```diff
@@ -0,0 +1,68 @@
+      out << ", \"adoptable\": true";
+    }
+    if (auto version = feature.getLanguageVersion()) {
+      out << ", \"introduced_in\": " << *version;
```
I used the phrasing from the Features.def but don’t have a strong preference.

xedin replied:
It wasn't in Features.def it's in Features.h for `Feature::getLanguageVersion()`.

xedin replied:
Changed to `enabled_in` and `adoptable` -> `migratable`

AnthonyLatsis replied:
cc @hborla 

Migratable was my suggestion.

xedin commented:
I added a test for `-print-supported-features` but it got lost among trying to get `-emit-supported-features` work, will bring it back tomorrow.

xedin commented on lib/Basic/SupportedFeatures.cpp:
```diff
@@ -0,0 +1,68 @@
+/// The information includes whether a feature is adoptable and for
+/// upcoming features - what is the first mode it's introduced.
+void printSupportedFeatures(llvm::raw_ostream &out) {
+  std::vector<swift::Feature> upcoming{{
```
I don't actually think it's worth it dealing with sizes here, that's why I used std::vector.

AnthonyLatsis replied:
The size should be deduced along with the element type if you omit the template arguments altogether. 

Not insisting! This was just me in pedantic mode knowing I had other comments anyway 😄

xedin replied:
✅

xedin commented on include/swift/Option/Options.td:
```diff
@@ -1534,6 +1534,10 @@ def print_target_info : Flag<["-"], "print-target-info">,
 
+def print_supported_features : Flag<["-"], "print-supported-features">,
+    Flags<[FrontendOption]>,
+    HelpText<"Print information about features supported by the compiler">, MetaVarName<"<features>">;
```
✅

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Linux platform

xedin committed:
[Frontend] Rework `-emit-supported-features` mode into `-emit-supported-arguments`

The "featues" part was never actually implemented and Swift Driver
is replying on information about arguments, so instead of removing
this mode, let's scope it down to "arguments" to be deprecated in
the future.

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
https://github.com/swiftlang/swift-driver/pull/1875
@swift-ci please test

xedin commented:
@swift-ci please test
